dali/base/dautils.cpp

```diff
@@ -783,94 +784,85 @@ bool CDfsLogicalFileName::normalizeExternal(const char * name, StringAttr &res,
+            // be calling setAllowWild(true), but maintaining current semantics which has always allowed
+            // wildcards because scopes were not being validated beyond this point before (HPCC-28885)
+            allowWild = true;
+            break;
         }
-        else
+        case lfntype_plane:
         {
-            str.append(s);
-            str.toLowerCase();
-        }
-        res.set(str);
-        return true;
-    }
+            //Syntax plane::<plane>::<path>
 
-    if (startsWithIgnoreCase(name, PLANE_SCOPE "::"))
-    {
-        //Syntax plane::<plane>::<path>
-        lfn.clear();
-        StringBuffer str;
-        const char *s=strstr(name,"::");
-        normalizeScope(name, name, s-name, str, strict);    // "plane"
+            StringBuffer planeName;
+            normalizeScope(s1, s1, ns1-s1, planeName, strict);
 
-        //find the name of the plane
-        const char *s1 = s+2;
-        const char *ns1 = strstr(s1,"::");
-        if (!ns1)
-            return false;
+            str.append("::").append(planeName);
+            break;
+        }
+        case lfntype_remote:
+        {
+            //Syntax plane::<remote>::<path>
 
-        StringBuffer planeName;
-        normalizeScope(s1, s1, ns1-s1, planeName, strict);
+            StringBuffer remoteSvc;
+            normalizeScope(s1, s1, ns1-s1, remoteSvc, strict);
 
-        Owned<IStoragePlane> plane = getDataStoragePlane(planeName, true);
-        if (plane->numDevices() == 0)
-            throw makeStringExceptionV(-1, "Scope contains invalid storage plane '%s'", planeName.str());
-            
-        str.append("::").append(planeName);
-        str.append(ns1);
-        str.toLowerCase();
-        res.set(str);
-        return true;
+            str.append("::").append(remoteSvc);
+            break;
+        }
     }
 
-    if (startsWithIgnoreCase(name, REMOTE_SCOPE "::"))
-    {
-        //Syntax plane::<remote>::<path>
-        lfn.clear();
-        StringBuffer str;
-        const char *s=strstr(name,"::");
-        normalizeScope(name, name, s-name, str, strict);    // "remote"
-
-        //find the name of the remote (service)
-        const char *s1 = s+2;
-        const char *ns1 = strstr(s1,"::");
-        if (!ns1)
-            return false;
-
-        StringBuffer remoteSvc;
-        normalizeScope(s1, s1, ns1-s1, remoteSvc, strict);
-
-        str.append("::").append(remoteSvc);
-        str.append(ns1);
-        str.toLowerCase();
-        res.set(str);
-        return true;
-    }
+    str.toLowerCase();
+    str.append("::");
+    unsigned len = str.length(); // length of <file|plane|remote>::<name>::
+    // handle trailing scopes+name
+    StringAttr tail;
+    normalizeName(ns1+2, tail, strict, false); // +2 skipping "::", validated at start
+    // normalizeName sets tailpos relative to ns1+2
+    tailpos += len; // need to add len. NB: tailpos points to the last scope, which generally is the file name
```

Comment:
How about call: tailpos += str.length(); without using len? 
Please correct me if I am wrong: when we check the scope access for a folder, the last scope could be a folder name. 
I found that the normalizeName() failed when the ns1 is '::folder::'.

Reply:
>the last scope could be a folder name.

Yes, it could be a scope (not a filename), it's most commonly used to point to a scoped filename though.

>I found that the normalizeName() failed when the ns1 is '::folder::'.

a trailing :: is invalid.

Reply:
>How about call: tailpos += str.length(); without using len?

could be, I'll change it.

Reply:
> a trailing :: is invalid.

Should the caller to the normalizeExternal() check and, if any, remove the tailing :: (or path separator)?

Reply:
The FileList method in the ESP FileSpray servise () calls the getDropZoneScopePermissions() to check the scope access to a LZ folder. The relative path may have a trailing path separator (such as /foldername/). The relative path is given to the setPlaneExternal() which, I think, will create a trailing :: (such as '::foldername::').

Reply:
It was a bug. New JIRA: https://track.hpccsystems.com/browse/HPCC-28891